File: backend/app/services/auth_service.py
"""
Authentication Service
Handles password hashing, JWT token generation, and user management
"""

from datetime import datetime, timedelta
from typing import Optional
from jose import JWTError, jwt
import bcrypt
from sqlalchemy.orm import Session
from sqlalchemy.exc import IntegrityError

from app.models.user import User
from app.config import settings


class AuthService:
    """Service for authentication operations"""
    
    @staticmethod
    def hash_password(password: str) -> str:
        """
        Hash a plain text password using bcrypt
        
        Args:
            password: Plain text password (must be ≤72 bytes UTF-8)
            
        Returns:
            Bcrypt hash as string
        """
        # Encode password to bytes (UTF-8)
        password_bytes = password.encode('utf-8')
        
        # Generate salt and hash
        salt = bcrypt.gensalt(rounds=12)
        hashed = bcrypt.hashpw(password_bytes, salt)
        
        # Return hash as string
        return hashed.decode('utf-8')
    
    @staticmethod
    def verify_password(plain_password: str, hashed_password: str) -> bool:
        """
        Verify a plain password against its bcrypt hash
        
        Args:
            plain_password: Plain text password
            hashed_password: Bcrypt hash string
            
        Returns:
            True if password matches, False otherwise
        """
        # Encode both to bytes
        password_bytes = plain_password.encode('utf-8')
        hash_bytes = hashed_password.encode('utf-8')
        
        # Verify using bcrypt
        return bcrypt.checkpw(password_bytes, hash_bytes)
    
    @staticmethod
    def create_access_token(data: dict, expires_delta: Optional[timedelta] = None) -> str:
        """
        Create a JWT access token
        
        Args:
            data: Dictionary with claims to encode in the token (typically {"sub": user_id})
            expires_delta: Optional custom expiration time
            
        Returns:
            Encoded JWT token string
        """
        to_encode = data.copy()
        
        if expires_delta:
            expire = datetime.utcnow() + expires_delta
        else:
            expire = datetime.utcnow() + timedelta(minutes=settings.ACCESS_TOKEN_EXPIRE_MINUTES)
        
        to_encode.update({"exp": expire})
        encoded_jwt = jwt.encode(to_encode, settings.SECRET_KEY, algorithm=settings.ALGORITHM)
        
        return encoded_jwt
    
    @staticmethod
    def decode_access_token(token: str) -> Optional[dict]:
        """
        Decode and validate a JWT token
        
        Args:
            token: JWT token string
            
        Returns:
            Decoded token payload or None if invalid
        """
        try:
            payload = jwt.decode(token, settings.SECRET_KEY, algorithms=[settings.ALGORITHM])
            return payload
        except JWTError:
            return None
    
    @staticmethod
    def create_user(db: Session, email: str, password: str) -> User:
        """
        Create a new user with hashed password
        
        Args:
            db: Database session
            email: User email (must be unique)
            password: Plain text password
            
        Returns:
            Created User object
            
        Raises:
            IntegrityError: If email already exists
        """
        hashed_password = AuthService.hash_password(password)
        
        user = User(
            email=email,
            hashed_password=hashed_password,
            is_active=True
        )
        
        db.add(user)
        db.commit()
        db.refresh(user)
        
        return user
    
    @staticmethod
    def get_user_by_email(db: Session, email: str) -> Optional[User]:
        """
        Get user by email address
        
        Args:
            db: Database session
            email: User email to search for
            
        Returns:
            User object or None if not found
        """
        return db.query(User).filter(User.email == email).first()
    
    @staticmethod
    def get_user_by_id(db: Session, user_id: str) -> Optional[User]:
        """
        Get user by UUID
        
        Args:
            db: Database session
            user_id: User UUID string
            
        Returns:
            User object or None if not found
        """
        return db.query(User).filter(User.id == user_id).first()
    
    @staticmethod
    def authenticate_user(db: Session, email: str, password: str) -> Optional[User]:
        """
        Authenticate user with email and password
        
        Args:
            db: Database session
            email: User email
            password: Plain text password
            
        Returns:
            User object if authentication successful, None otherwise
        """
        user = AuthService.get_user_by_email(db, email)
        
        if not user:
            return None
        
        if not AuthService.verify_password(password, user.hashed_password):
            return None
        
        if not user.is_active:
            return None
        
        return user

File: backend/src/services/AuthService.ts
import bcryptjs from 'bcryptjs';
import jwt from 'jsonwebtoken';
import { User } from '../models/index.js';
import { config } from '../config/environment.js';
import { AuthenticationError, ValidationError } from '../utils/errors.js';

export class AuthService {
  static async createUser(email: string, password: string) {
    // Validate email and password
    if (!email || !password) {
      throw new ValidationError('Email and password are required');
    }

    if (password.length < 8) {
      throw new ValidationError('Password must be at least 8 characters');
    }

    // Check if user already exists
    const existingUser = await User.findOne({ email: email.toLowerCase() });
    if (existingUser) {
      throw new ValidationError('Email already registered');
    }

    // Hash password
    const hashedPassword = await bcryptjs.hash(password, 12);

    // Create user
    const user = await User.create({
      email: email.toLowerCase(),
      hashedPassword,
    });

    return user;
  }

  static async authenticateUser(email: string, password: string) {
    const user = await User.findOne({ email: email.toLowerCase() });

    if (!user) {
      throw new AuthenticationError('Incorrect email or password');
    }

    const isPasswordValid = await bcryptjs.compare(password, user.hashedPassword);

    if (!isPasswordValid) {
      throw new AuthenticationError('Incorrect email or password');
    }

    if (!user.isActive) {
      throw new AuthenticationError('User account is inactive');
    }

    return user;
  }

  static async getUserById(userId: string) {
    const user = await User.findById(userId).select('-hashedPassword');
    return user;
  }

  static async getUserByEmail(email: string) {
    const user = await User.findOne({ email: email.toLowerCase() }).select('-hashedPassword');
    return user;
  }

  static createAccessToken(userId: string): string {
    const token = jwt.sign({ sub: userId }, config.JWT_SECRET, {
      expiresIn: config.JWT_EXPIRE_IN,
    });
    return token;
  }

  static verifyToken(token: string): { sub: string } {
    const decoded = jwt.verify(token, config.JWT_SECRET) as { sub: string };
    return decoded;
  }

  static async changePassword(userId: string, oldPassword: string, newPassword: string) {
    if (!newPassword || newPassword.length < 8) {
      throw new ValidationError('Password must be at least 8 characters');
    }

    const user = await User.findById(userId);
    if (!user) {
      throw new AuthenticationError('User not found');
    }

    const isOldPasswordValid = await bcryptjs.compare(oldPassword, user.hashedPassword);
    if (!isOldPasswordValid) {
      throw new AuthenticationError('Current password is incorrect');
    }

    const hashedPassword = await bcryptjs.hash(newPassword, 12);
    user.hashedPassword = hashedPassword;
    await user.save();

    return user;
  }
}

export default AuthService;
